posts/benchmarking_design_decisions/index.qmd

```diff
@@ -0,0 +1,246 @@
+
+```{r}
+library(bench)
+library(cli)
+library(ggplot2)
+library(purrr)
+library(dplyr)
+library(tidyr)
+```
+
+Let's define a function that takes an argument `n` for the number of warnings
+to throw and `pkg` for the type of package to use.
+
+```{r}
+warn_Ntimes <- function(n, pkg) {
+  warning_msg <- "x must be an integer"
+  switch(pkg,
+    base = for (i in 1:n) {
+      warning(warning_msg)
+    },
+    cli = for (i in 1:n) {
+      cli_warn(warning_msg)
+    }
+  )
+}
+```
+
+Let's test our function to see if it works as expected.
+
+```{r}
+warn_Ntimes(3, "base")
+warn_Ntimes(3, "cli")
+```
+
+Now, we'll consider scenarios where a function throws 1, 5, 10, 15, 20, and
+100 warnings using base R and `cli`.
+
+```{r warning=FALSE}
+# Number of warnings to throw
+warnings <- c(1, 5, 10, 15, 20, 100)
+```
+
+Let's run benchmarks over the different scenarios and store results
+in a data.frame.
+
+```{r warning=FALSE,message=FALSE}
+bnmark_res <- bench::press(
+  warnings = c(1, 5, 10, 15, 20, 100),
+  bench::mark(
+    cli = warn_Ntimes(warnings, "cli"),
+    base = warn_Ntimes(warnings, "base")
+    )
+  )
+```
+
+## Results
+
+It's time to explore the results of the data generated. We'll start by making
+a boxplot of the median run times for the different scenarios.
+
+Drum roll please...
+
+```{r}
+autoplot(
+  bnmark_res,
+  type = "boxplot"
+  ) +
+  labs(
+    x = "Package",
+    caption = "Functions throwing various numbers of warnings as indicated in the facet label"
+  )
+```
+
+Finally, let's investigate the relative speeds of the two packages with
+respect to the number of warnings. We will do this by dividing the median
+benchmark of `cli::cli_warn()` by that of `base::warning()`.
+
+```{r}
+bnmark_relative_speeds <- bnmark_res |>
+  select(expression, median, warnings) |>
+  pivot_wider(
+    names_from = expression,
+    values_from = c(median)
+    ) |> 
+  mutate(
+    relative_speed = cli/base
+    )
+bnmark_relative_speeds
+```
+
+As we can see in the new column `relative_speed`, `cli` is consistently
+slower than base R and this seems not to
+scale linearly with the number of warnings thrown. Other benchmarking
+experiments
+with `cli` have revealed linear relationships between its speed and the number
+of times it invokes other functions [^3]. The developers of `cli` have also
+conducted benchmarks of the `ansi_` family of functions in `cli` in comparison
+to base R and the `fansi` package. They find that `cli` is consistently slower
+than base R, which corroborates the results of our experiment here.
+Their benchmarks are available in the cli documentation [^4]. So, should we
+be worried about the speed of `cli`? Well, it depends on the context. The
+"R Packages" book by Hadley Wickham and Jenny Bryan has a section on
+this where they suggest approaching such a decision from a holistic, balanced,
+and quantitative approach[^5]. We'll leave the reader to make their own decision
```

Comment:
It might be worth fully citing the R packages book here. I don't know if we have a policy/preference on citations vs footnotes. I think citing would be a more clear sign of attribution for a published piece of work.

@Bisaloo any preferences?

Reply:
That's true.

Reply:
I've switched to using citations for books.